options/ansi/generic/time-stubs.cpp

```diff
@@ -578,10 +579,445 @@ char *ctime_r(const time_t *clock, char *buf) {
+    *string_index = si;
+    return count;
+}
+
+int month_to_day(int month) {
+    switch(month){
+        case  0: return 0;
+        case  1: return 31;
+        case  2: return 59;
+        case  3: return 90;
+        case  4: return 120;
+        case  5: return 151;
+        case  6: return 181;
+        case  7: return 212;
+        case  8: return 243;
+        case  9: return 273;
+        case 10: return 304;
+        case 11: return 334;
+    }
+    return -1;
+}
+
+int is_leapyear(int year) {
+    return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
+}
+
+int month_and_year_to_day_in_year(int month, int year){
+    int day = month_to_day(month);
+    if(is_leapyear(year) && day < month_to_day(2)) //TODO just do 2 instead of month_to_day(2) lol
+        return day + 1;
+
+    return day;
+}
+
+int target_determination(int month) {
+    switch(month){
+        case 0: return 3;
+        case 1: return 14;
+        case 2: return 14;
+        case 3: return 4;
+        case 4: return 9;
+        case 5: return 6;
+        case 6: return 11;
+        case 7: return 8;
+        case 8: return 5;
+        case 9: return 10;
+        case 10: return 7;
+        case 11: return 12;
+    }
+
+    return -1;
+}
+
+int doom_determination(int full_year) {
+    int century = full_year / 100;
+    int anchor = 2 + 5 * (century % 4) % 7;
+
+    int year = full_year % 100;
+
+    if(year % 2)
+        year += 11;
+
+    year /= 2;
+
+    if(year % 2)
+        year += 11;
+
+    return 7 - (year % 7) + anchor;
+}
+
+//determine day of week through the doomsday algorithm,
+//invented by someone much smarter than me (Conway).
+int day_determination(int day, int month, int year) {
+    int doom = doom_determination(year);
+    bool leap = is_leapyear(year);
+
+    int target = target_determination(month);
+    if(leap && month < 2)
+        target++;
+
+    int doom_dif = (day - target) % 7;
+    return (doom + doom_dif) % 7;
+}
+
+struct strptime_internal_state {
+    bool has_century;
+    bool has_year;
+    bool has_month;
+    bool has_day_of_month;
+    bool has_day_of_year;
+    bool has_day_of_week;
+
+    bool full_year_given;
+
+    int century;
+
+    size_t format_index;
+    size_t input_index;
+};
+
+char *strptime_internal(const char *__restrict input, const char *__restrict format,
+	struct tm *__restrict tm, struct strptime_internal_state *__restrict state) {
+    while(isspace(input[state->input_index])) state->input_index++;
+    
+    if(input[state->input_index] == '\0')
+        return NULL;
+
+    while(format[state->format_index] != '\0'){
+        if(format[state->format_index] == '%'){
+            state->format_index++;
+            switch(format[state->format_index]){
+                case '%':
+                    if(input[state->input_index++] != '%')
+                        return NULL;
+                    break;
+                case 'a':
+                case 'A':
+                    break;
```

Comment:
If there are any unimplemented cases, they should at least be logged, if not just fail the function.

Reply:
Unless I missed any, it should now `__ensure()` that unimplemented cases are not reached.